### PR TITLE
Pass memory/cpu limits as env variables to dynamic service

### PR DESCRIPTION
**Pull request by mguidon**
## What do these changes do?

Pass the cpu/memory limits to the running containers (comp. and dynamic) so that the guest application can consider them.

## Related issue/s

https://git.speag.com/oSparc/osparc-s4l/-/issues/151

## How to test

Create services (comp/dyn) and check env vars assigned to them.

### CODE REVIEW

**Review by GitHK — CHANGES_REQUESTED** [review]
I would avoid multiple redefinitions of:
```python
CPU_RESOURCE_LIMIT_KEY = "SIMCORE_NANO_CPUS_LIMIT" 
MEM_RESOURCE_LIMIT_KIT = "SIMCORE_MEMORY_BYTES_LIMIT"
```

This will be used in the sidecar, and dynamic-sidecar and director-v2 ( last 2 coming via #1887).

Please consider moving the definitions to the `service-library`.

@sanderegg anything agains this?

File: services/sidecar/src/simcore_service_sidecar/config.py [hunk 1]
```diff
@@ -83,3 +83,7 @@
 REDLOCK_REFRESH_INTERVAL_SECONDS: float = max(
     float(os.environ.get("REDLOCK_REFRESH_INTERVAL_SECONDS", "5.0")), 1.0
 )  # enforce at least 1 second
+
+# resources
+CPU_RESOURCE_LIMIT_KEY = "SIMCORE_NANO_CPUS_LIMIT"
+MEM_RESOURCE_LIMIT_KIT = "SIMCORE_MEMORY_BYTES_LIMIT"
```
[the remark this answers is not in the document]
Comment (mguidon) [reply services/sidecar/src/simcore_service_sidecar/config.py:89]: And where should I put them?

[the remark this answers is not in the document]
Comment (GitHK) [reply services/sidecar/src/simcore_service_sidecar/config.py:89]: Thought I've already replied to this. Inside `packages/service-library/src/servicelib/resources.py`

File: services/director/src/simcore_service_director/producer.py [hunk 2]
```diff
@@ -318,12 +322,21 @@ async def _create_docker_service_params(
...
+    mem_limit = str(
         docker_params["task_template"]["Resources"]["Limits"]["MemoryBytes"]
     )
+    container_spec["Labels"]["nano_cpus_limit"] = nano_cpus_limit
+    container_spec["Labels"]["mem_limit"] = mem_limit
+
+    # and make the container aware of them via env variables
+    resource_limits = {
+        CPU_RESOURCE_LIMIT_KEY: nano_cpus_limit,
+        MEM_RESOURCE_LIMIT_KIT: mem_limit,
```
Comment (sanderegg) [inline services/director/src/simcore_service_director/producer.py:337]:
```suggestion
        MEM_RESOURCE_LIMIT_KEY: mem_limit,
```

File: services/director/tests/test_producer.py [hunk 3]
```diff
@@ -172,6 +171,9 @@ async def test_service_assigned_env_variables(run_services, user_id, project_id)
         assert "SIMCORE_HOST_NAME" in envs_dict
         assert envs_dict["SIMCORE_HOST_NAME"] == docker_service.name
 
+        assert config.MEM_RESOURCE_LIMIT_KIT in envs_dict
```
Comment (sanderegg) [inline services/director/tests/test_producer.py:174]:
```suggestion
        assert config.MEM_RESOURCE_LIMIT_KEY in envs_dict
```

File: services/sidecar/src/simcore_service_sidecar/config.py (hunk 1 above)
Comment (sanderegg) [inline services/sidecar/src/simcore_service_sidecar/config.py:89]:
```suggestion
MEM_RESOURCE_LIMIT_KEY = "SIMCORE_MEMORY_BYTES_LIMIT"
```

**Review by sanderegg — APPROVED** [review]
Some changes to the const names. besides that I agree with @GitHK .

File: services/director/src/simcore_service_director/config.py [hunk 4]
```diff
@@ -125,4 +125,8 @@
     "TRACING_ZIPKIN_ENDPOINT", "http://jaeger:9411"
 )
 
+# resources
+CPU_RESOURCE_LIMIT_KEY = "SIMCORE_NANO_CPUS_LIMIT"
+MEM_RESOURCE_LIMIT_KIT = "SIMCORE_MEMORY_BYTES_LIMIT"
+
```
[the remark this answers is not in the document]
Comment (mguidon) [reply services/director/src/simcore_service_director/config.py]: Not possible for the director (installs specific hash for the service lib)

File: services/director/src/simcore_service_director/producer.py (hunk 2 above)
Comment (mguidon) [reply services/director/src/simcore_service_director/producer.py]: Done

File: services/sidecar/src/simcore_service_sidecar/config.py (hunk 1 above)
Comment (mguidon) [reply services/sidecar/src/simcore_service_sidecar/config.py]: Done

[the remark this answers is not in the document]
Comment (mguidon) [reply services/sidecar/src/simcore_service_sidecar/config.py]: Done for sidecar

File: services/director/src/simcore_service_director/config.py [hunk 5]
```diff
@@ -125,4 +125,8 @@
     "TRACING_ZIPKIN_ENDPOINT", "http://jaeger:9411"
 )
 
+# resources
+CPU_RESOURCE_LIMIT_KEY = "SIMCORE_NANO_CPUS_LIMIT"
+MEM_RESOURCE_LIMIT_KEY = "SIMCORE_MEMORY_BYTES_LIMIT"
+
```
Comment (GitHK) [inline services/director/src/simcore_service_director/config.py:131]: do not forget to remove these

File: services/director/src/simcore_service_director/producer.py [hunk 6]
```diff
@@ -17,7 +17,11 @@
 from servicelib.monitor_services import service_started, service_stopped
 
 from . import config, docker_utils, exceptions, registry_proxy
-from .config import APP_CLIENT_SESSION_KEY
+from .config import (
+    APP_CLIENT_SESSION_KEY,
+    CPU_RESOURCE_LIMIT_KEY,
+    MEM_RESOURCE_LIMIT_KEY,
+)
```
Comment (GitHK) [inline services/director/src/simcore_service_director/producer.py:24]: you will need to change the import here

File: services/director/src/simcore_service_director/config.py (hunk 5 above)
Comment (mguidon) [reply services/director/src/simcore_service_director/config.py]: I would if I could but I can't so I shan't.

**Review by GitHK — APPROVED** [review]

**Pull request merged**